paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala

```diff
@@ -0,0 +1,187 @@
+    }
+
+    // append _bucket_ column as placeholder
+    val withBucketCol = partitioned.withColumn(BUCKET_COL, lit(-1))
+    val bucketColIdx = withBucketCol.schema.size - 1
+
+    val dataEncoder = RowEncoder.apply(withBucketCol.schema).resolveAndBind()
+    val toRow = dataEncoder.createSerializer()
+    val fromRow = dataEncoder.createDeserializer()
+
+    val bucketProcessor = if (isDynamicBucketTable) {
+      val fileStoreTable = table.asInstanceOf[FileStoreTable]
+      DynamicBucketProcessor(fileStoreTable, rowType, bucketColIdx, toRow, fromRow)
+    } else {
+      CommonBucketProcessor(writeBuilder, bucketColIdx, toRow, fromRow)
+    }
+    val committables =
+      withBucketCol
+        .mapPartitions(bucketProcessor.processPartition)(dataEncoder)
+        .groupByKey(row => row.getInt(bucketColIdx))
+        .flatMapGroups {
+          (bucket, iter) =>
+            val write = writeBuilder.newWrite()
+            try {
+              val serializer = new CommitMessageSerializer
+              iter.foreach {
+                row =>
+                  // Notice: `row` contains a _bucket_ column, but there's no impact and we don't have to remove it.
+                  write.write(new DynamicBucketRow(new SparkRow(rowType, row), bucket))
+              }
+              write.prepareCommit().asScala.map(serializer.serialize)
+            } finally {
+              write.close()
+            }
+        }
+        .collect()
+        .map(deserializeCommitMessage(serializer, _))
+
+    try {
+      val tableCommit =
+        writeBuilder.newCommit().asInstanceOf[InnerTableCommit].withLock(lockFactory.create())
+      tableCommit.commit(committables.toList.asJava)
+    } catch {
+      case e: Throwable => throw new RuntimeException(e);
+    }
+
+    Seq.empty
+  }
+
+  override def withNewChildrenInternal(newChildren: IndexedSeq[LogicalPlan]): LogicalPlan =
+    this.asInstanceOf[WriteIntoPaimonTable]
+}
+
+object WriteIntoPaimonTable {
+
+  sealed trait BucketProcessor {
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row]
+  }
+
+  case class CommonBucketProcessor(
+      writeBuilder: BatchWriteBuilder,
+      bucketColIndex: Int,
+      toRow: ExpressionEncoder.Serializer[Row],
+      fromRow: ExpressionEncoder.Deserializer[Row])
+    extends BucketProcessor {
+
+    private val rowType = writeBuilder.rowType
+
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row] = {
+      val batchTableWrite = writeBuilder.newWrite()
+      new Iterator[Row] {
+        override def hasNext: Boolean = rowIterator.hasNext
+
+        override def next(): Row = {
+          val row = rowIterator.next
+          val sparkInternalRow = toRow(row)
+          sparkInternalRow.setInt(
+            bucketColIndex,
+            batchTableWrite.getBucket(new SparkRow(rowType, row)))
+          fromRow(sparkInternalRow)
+        }
+      }
+    }
+  }
+
+  case class DynamicBucketProcessor(
```

Comment:
DynamicBucketProcessor, different tasks will go to fill the bucket data, which may lead to a bucket assign too much data.

And it may lead to assign error, for example, the same key in different tasks, was assign different bucket.